docs/data/data-grid/pagination/pagination.md

```diff
@@ -122,7 +150,96 @@ React.useEffect(() => {
 
 {{"demo": "ServerPaginationGrid.js", "bg": "inline"}}
 
-### Cursor implementation
+#### Unknown row count
+
+When the row count is not available, you can use the following steps to enable server-side pagination:
+
+- Set the prop `paginationMode` to `server`
+- Set the `rowCount` as `-1` using one of the above mentioned ways to let the Grid know that the row count is not available
+- Since the `rowCount` is unknown, the Grid needs to know whether more records are available on the server. Pass the `paginationMeta.hasNextPage` boolean property to indicate whether more records are available.
+- Use the `onPaginationModelChange` prop to react to the `paginationModel` changes
+
+The `hasNextPage` must not be set to `false` until there are _actually_ no records left to fetch, because when `hasNextPage` becomes `false`, the Grid considers this as the last page and tries to set the `rowCount` value to a finite value.
+
+Relying on this behavior of the Grid should be the last resort though, the preferred way is to set the `rowCount` to the actual number of rows fetched from the server as soon as the information is available since the Grid cannot guess if the rows in the last page are less than the page size, and the incorrect number might be displayed in the pagination UI.
+
+For example the if the row count is `15` and the page size is `10`, on the last page, when `hasNextPage` becomes false the Grid will set the row count to be `page (2) * pageSize (10) = (20)` showing `10-20 of 20` which is incorrect, the correct value should be `10-15 of 15`.
```

Comment:
Do we really need this warning?
I think this is acceptable behavior when the row count is unknown, we don't have to discourage the users to do this if this is their use case.

One thing to clarify here – the incorrect number might appear only while fetching of the last page, and once the page is fetched – we update it with the actual number of rows (all of them have been fetched). Is this correct?

Reply:
> the incorrect number might appear only while fetching of the last page, and once the page is fetched – we update it with the actual number of rows (all of them have been fetched). Is this correct?

@cherniavskii Now, yes (done in https://github.com/mui/mui-x/pull/12474/commits/4f8f45dff590622d1284677a46c941c9cb230c5d), thanks for pointing it out.

Previously I was aligned to off-load this responsibility to the user (that they explicitly have to set the updated `rowCount` prop especially if the last page size on server is less that `paginationModel.pageSIze`), because loading could be set to `false` in one render cycle while rows might be updated in multiple render cycles or at a later point in time depending on how the user is fetching them, so it didn't seem safe to make it a default behavior.

However, on a rethought, it seemed fine to me (as well as a better DX) since it could be opted out by controlling the `rowCount` prop.

Removed the warning as well, since the behavior is pretty much self-explanatory to me now. Let me know if you feel like adding a couple of line may still help.